### PR TITLE
Implement DeadCom Layer 2, migrate docs

### PR DESCRIPTION
**Pull request by adam-dej**

### CODE REVIEW

File: dcl2/lib/inc/dcl2.h [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+ *
+ * This function returns the length of received message, if any.
+ *
+ * @param[in] deadcom  Instance of an open DeadCom link
+ *
+ * @return Length of the received message in bytes or
+ *         DC_E_NOMSG if no message is waiting or
+ *         DC_E_FAIL  if parameters are invalid
+ *
+ * @note   Since "dcGetReceivedMsg" and "dcGetReceivedMsg" are 2 different functions, one might
```
Comment (lubo) [inline dcl2/lib/inc/dcl2.h]: Did you mean *"dcGetReceivedMsg"* and *"dcGetReceivedMsgLen"*?

File: dcl2/lib/inc/dcl2.h [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+ *           - Return DC_E_NOMSG if link was reset in the meantime
+ *         In particular, this means that after dcGetReceivedMsgLen returns a number of bytes,
+ *         a buffer of that size is guaranteed to be sufficient to copy the received message.
+ */
+int16_t dcGetReceivedMsgLen(DeadcomL2 *deadcom);
+
+/**
+ * Get the received message.
+ *
+ * This function returns the received message, in any. When the message is copied this function
```
Comment (lubo) [inline dcl2/lib/inc/dcl2.h]: ***if** any*?

File: dcl2/lib/src/dcl2.c [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                    if (yahdlc_frame_data(&resp_ctrl, NULL, 0, NULL, &f_len) == -EINVAL) {
+                        deadcom->t->mutexUnlock(deadcom->mutex_p);
+                        return DC_FAILURE;
+                    }
+
+                    {
+                        uint8_t resp_f[f_len];
+                        yahdlc_frame_data(&resp_ctrl, NULL, 0, resp_f, &f_len);
+                        deadcom->transmitBytes(resp_f, f_len);
+                    }
```
Comment (lubo) [inline dcl2/lib/src/dcl2.c]: Why did you put this block into braces?

File: docs/dcl2/protocol.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
+Deadcom Layer 2 Specification
+=============================
+
+Goal of this protocol is to provide a reliable datagram communication between
+exactly two devices connected using a point-to-point link.
+
+This protocol was heavily inspired by [HDLC (High-Level Data Link Control)](https://en.wikipedia.org/wiki/High-Level_Data_Link_Control), but simplified for our
```
Comment (lubo) [inline docs/dcl2/protocol.md:7]: How about using defining a block-level hyperlink target?

File: test/dcl2-integration/common.c [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    DeadcomL2 *station;
+    leaky_pipe_t *rx_pipe;
+    char station_char;
+} rx_set_t;
+
+void* rx_handle_thread(void *p) {
+    rx_set_t *rp = (rx_set_t*) p;
+    uint8_t b[1];
+    while (lp_receive(rp->rx_pipe, b, 1)) {
+        // printf("Station %c received %02x\n", rp->station_char, b[0]);
```
Comment (lubo) [inline test/dcl2-integration/common.c]: Nobody likes code that is commented out without a obvious reason (at least it's not obvious to me).

**Review by lubo — COMMENTED** [review]
If you did create a separate repository for your yahdlc fork, it'd be easier to see your changes.

File: dcl2/lib/src/dcl2.c (hunk 3 above)
Comment (adam-dej) [reply dcl2/lib/src/dcl2.c]: `resp_f` is an identiﬁer having a variably modiﬁed type, and the following `case` statement on line 358 would jump to its scope, which is prohibited by the C standard. Those braces create an explicit scope for `resp_f`.

File: test/dcl2-integration/common.c (hunk 5 above)
Comment (adam-dej) [reply test/dcl2-integration/common.c]: I've added a comment explaining what it may be useful for.

File: dcl2/lib/inc/dcl2.h (hunk 2 above)
Comment (adam-dej) [reply dcl2/lib/inc/dcl2.h]: Fixed

File: dcl2/lib/inc/dcl2.h (hunk 1 above)
Comment (adam-dej) [reply dcl2/lib/inc/dcl2.h]: Fixed

File: docs/dcl2/protocol.md (hunk 4 above)
Comment (adam-dej) [reply docs/dcl2/protocol.md:7]: I'm sorry, I'm not sure what you mean by a block-level hyperlink target (in markdown).

Comment (lubo) [reply docs/dcl2/protocol.md:7]: Oh, I didn't noticed this is markdown. My bad, ignore my comment.

**Comment by adam-dej** [comment]
This PR is no longer WIP, please perform a second round of review.

**Review by lubo — APPROVED** [review]

**Pull request merged**